pkg/pcr/flow_detection.go

```diff
@@ -1,13 +1,96 @@
 package pcr
 
 import (
+	"bytes"
 	"fmt"
+	"github.com/9elements/converged-security-suite/v2/pkg/intel/metadata/manifest"
 
 	"github.com/9elements/converged-security-suite/v2/pkg/errors"
 	"github.com/9elements/converged-security-suite/v2/pkg/intel/metadata/fit"
 	"github.com/9elements/converged-security-suite/v2/pkg/registers"
 )
 
+func DetectTPM(firmware Firmware, regs registers.Registers) (registers.TPMType, error) {
+	// We have two approaches:
+	// - based on registers provides a reliable results, but these values may not exist
+	// - based on firmware may provide hints that TPM2.0 is not supported
+	btgACMInfo, found := registers.FindBTGSACMInfo(regs)
+	if found {
+		return btgACMInfo.TPMType(), nil
+	}
+
+	acmPolicyStatus, found := registers.FindACMPolicyStatus(regs)
+	if found {
+		return acmPolicyStatus.TPMType(), nil
+	}
+
+	if firmware != nil {
+		fitEntries, err := fit.GetEntries(firmware.Buf())
+		if err != nil {
+			return 0, fmt.Errorf("unable to parse FIT entries: %w", err)
+		}
+
+		for _, entry := range fitEntries {
+			switch fitEntry := entry.(type) {
+			case *fit.EntrySACM:
+				data, err := fitEntry.ParseData()
+				if data == nil {
+					return 0, fmt.Errorf("unable to parse EntrySACM: %v", err)
+				}
+				_, chipset, err := manifest.ParseChipsetACModuleInformation(bytes.NewBuffer(data.UserArea))
+				if err != nil {
+					return 0, fmt.Errorf("failed to read ChipsetACModuleInformation, err: %v", err)
+				}
+
+				// From Intel TXT Software Development Guide:
+				// Version 5 included all
+				// changes added to support TPM 2.0 family.
+				if chipset.Base.Version < 5 {
+					return registers.TPMType12, nil
+				}
+
+				// chipset.TPMInfoList is an offset in bytes from ACM start.
+				image := firmware.ImageBytes()
+				var tpmInfo manifest.TPMInfoList
+				_, err = tpmInfo.ReadFrom(bytes.NewBuffer(image[fitEntry.GetDataOffset()+uint64(chipset.TPMInfoList):]))
+				if err != nil {
+					return 0, fmt.Errorf("failed to read TPMInfoList, err: %v", err)
+				}
+
+				bool2Int := func(b bool) int {
+					if b {
+						return 1
+					}
+					return 0
+				}
+				tpmFamilySupport := tpmInfo.Capabilities.TPMFamilySupport()
+				// if none options is set - no TPM
+				// if only one option is set - can determine
+				s := bool2Int(tpmFamilySupport.IsDiscreteTPM20Supported()) +
+					bool2Int(tpmFamilySupport.IsFirmwareTPM20Supported()) +
+					bool2Int(tpmFamilySupport.IsDiscreteTPM12Supported())
+
+				if s == 0 {
+					return registers.TPMTypeNoTpm, nil
+				}
+				if s == 1 {
```

Comment:
What should we do if `s > 1`?

Reply:
Suppose tpm2 by default...

I also have a PR where we get user-defined TPM value. Only if it is not set, this fallback detection will be used.
So, the PCR0 calc algorithm will take the following tuple: <required firmware, optional registers, optional tpm_version>.
If something is not provided, just try our best (like this fallback) with the data we have :)